docker-app/qfieldcloud/core/models.py

```diff
@@ -774,79 +775,17 @@ class RoleOrigins(models.TextChoices):
-                V(ProjectCollaborator.Roles.ADMIN),
-                V(ProjectQueryset.RoleOrigins.PROJECTOWNER),
-            ),
-            # Organization memberships - admin
-            (
-                Q(owner__in=Organization.objects.filter(organization_owner=user)),
-                V(ProjectCollaborator.Roles.ADMIN),
-                V(ProjectQueryset.RoleOrigins.ORGANIZATIONOWNER),
-            ),
-            (
-                Q(
-                    owner__in=OrganizationMember.objects.filter(
-                        member=user, role=OrganizationMember.Roles.ADMIN
-                    ).values("organization")
-                ),
-                V(ProjectCollaborator.Roles.ADMIN),
-                V(ProjectQueryset.RoleOrigins.ORGANIZATIONADMIN),
-            ),
-            # Role through ProjectCollaborator
-            (
-                Exists(
-                    ProjectCollaborator.objects.validated().filter(
-                        project=OuterRef("pk"),
-                        collaborator=user,
-                    )
-                ),
-                ProjectCollaborator.objects.filter(
-                    project=OuterRef("pk"),
-                    collaborator=user,
-                ).values_list("role"),
-                V(ProjectQueryset.RoleOrigins.COLLABORATOR),
-            ),
-            # Role through Team membership
-            (
-                Exists(
-                    ProjectCollaborator.objects.filter(
-                        project=OuterRef("pk"),
-                        collaborator__team__members__member=user,
-                    )
-                ),
-                ProjectCollaborator.objects.filter(
-                    project=OuterRef("pk"),
-                    collaborator__team__members__member=user,
-                ).values_list("role"),
-                V(ProjectQueryset.RoleOrigins.TEAMMEMBER),
-            ),
-            # Public
-            (
-                Q(is_public=True),
-                V(ProjectCollaborator.Roles.READER),
-                V(ProjectQueryset.RoleOrigins.PUBLIC),
-            ),
-        ]
-
-        qs = self.annotate(
-            user_role=Case(
-                *[When(perm[0], perm[1]) for perm in permissions_config],
-                default=None,
-                output_field=models.CharField(),
-            ),
-            user_role_origin=Case(
-                *[When(perm[0], perm[2]) for perm in permissions_config],
-                default=None,
-            ),
+        qs = (
+            Project.objects.select_related("user_role")
+            .defer("user_role__user_id", "user_role__project_id")
+            .annotate(user_role__user_id=Coalesce("user_role__user_id", None))
+            .filter(
+                Q(user_role__user=user)
+                | Q(user_role__origin=ProjectQueryset.RoleOrigins.PUBLIC),
+                user_role__is_valid=True,
+            )
+            .distinct("owner__username", "name")
```

Comment:
This distinct here is not ok. It prevents the `for_user` qs to be sorted, as it translates to `DISTINCT ON`. Ideas?

Reply:
Do the duplicates appear because an user can have several roles in the view ? If so, could it be sorted by having the DISTINCT clause in the view instead ?

Reply:
The problem comes with the public projects, as other rows in the view already have `DISTINCT ON (project_id, user_id)`, but the public projects will return `NULL` as user_id in the query. The more I think, the more I believe the public projects should be handled separately.

Reply:
Yes, in the API list view for instance we had to handle public projects differently than other roles already (because you do not want to include them in your default project's list). And the day we want to materialize the view (either just materialized view, or a regular django model updated through signals), we won't be able to keep a list of all public relations anyway.

I think it's quite clear : `Project.objects.filter( Q(project_roles__user=user1) | Q(is_public=True) )`

Reply:
not really, because we need to explude the duplicates

Reply:
In my example above, if the permissions view has unique rows for project, user, I think there should be no duplicate? The | operator on Q objects is supposed to resolve to an OR on the where clause (not to be confused with | on queryset which would be an union).
Or am I missing something?

Reply:
Ok, we did it! This delayed union is doing mirracles.

Reply:
well, miracles only in tests here, failed in the UI part. Anyway, lessons learned that `row_number` is very slow.